### DISCUSSION TITLE
JsonException: Control character error, possibly incorrectly encoded

### DISCUSSION THREAD

**Discussion by NickMous** [deployphp/deployer#3981] [Help needed] [answered]
Hiya there, I'm trying to set up a new project with deployer 8, and I'm getting a JsonException the whole time, without any useful information.
Can anyone help me debug this, if I'm the problem or the script is?

It's a fresh laravel application.

```
[<project-stag>] run cd /var/www/<project>/01 && (ln -nfs --relative releases/6 /var/www/<project>/01/release)
done on <project-stag>
  ⠼
In Server.php line 190:

  [JsonException (3)]
  Control character error, possibly incorrectly encoded


Exception trace:
  at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:190
 json_decode() at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:190
 Deployer\Executor\Server->parseRequest() at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:142
 Deployer\Executor\Server->handleClientRequests() at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:71
 Deployer\Executor\Server->run() at /var/www/html/vendor/deployer/deployer/src/Executor/Master.php:230
 Deployer\Executor\Master->runTask() at /var/www/html/vendor/deployer/deployer/src/Executor/Master.php:110
 Deployer\Executor\Master->run() at /var/www/html/vendor/deployer/deployer/src/Command/MainCommand.php:135
 Deployer\Command\MainCommand->execute() at /var/www/html/vendor/symfony/console/Command/Command.php:279
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/console/Application.php:1076
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:342
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/symfony/console/Application.php:193
 Symfony\Component\Console\Application->run() at /var/www/html/vendor/deployer/deployer/src/Deployer.php:289
 Deployer\Deployer::run() at /var/www/html/vendor/deployer/deployer/bin/dep:103
 include() at /var/www/html/vendor/bin/dep:119

deploy [--tag TAG] [--revision REVISION] [--branch BRANCH] [-o|--option OPTION] [-l|--limit LIMIT] [--no-hooks] [--plan] [--start-from START-FROM] [--log LOG] [--profile PROFILE] [--] [<selector>...]

Failed to execute command vendor/bin/dep deploy -vvv: exit status 3
```

**Comment by NickMous** [comment]
Did some more digging in Server.php, and it's not completely giving a correct array.
The code `array_slice($lines, $i + 1)`, which is in the json_encode (imploded), does not result in a correct array. See example:

```
array (
  0 => '{
    "host": "<proj-stag>",
    "config": {
        "alias": "<proj-stag>",
        "hostname": "<host>",
        "remote_user": "deployer",
        "port": 1124,
        "deploy_path": "\\/var\\/www\\/<proj-stag>",
        "selected_hosts": [
            "<proj-stag>"
        ],
        "shell": "bash -ls",
        "env": [],
        "default_timeout": 300,
        "dotenv": false,
        "forward_agent": true,
        "ssh_multiplexing": true,
        "repository": "git@github.com:NickMous\\/<proj>",
        "what": "zcdb",
        "where": "zcdb01",
        "working_path": null,
        "current_path": "\\/var\\/www\\/<proj>\\/current",
        "user": "DDEV User",
        "releases_list": [
            "1"
        ],
        "release_name": "1",
        "branch": "HEAD",
        "target": "HEAD",
        "use_relative_symlink": true,
        "bin\\/symlink": "ln -nfs --relative"
  ',
)
```

No, I did not cut this off.

**Comment by NickMous** [comment]
Found the issue! And it looks like it's not me...
In `src\Executor\Server.php::readClientRequest()` the socket is being read until we have 2 new lines.
Somehow in the socket there are two new lines, but it's not done reading.
I altered the code to output everything in separate files, and if people are interested I can provide it.
For now, I'm gonna create a bug issue and delve in it a bit deeper!

**Reply by vbahinwillit** [reply]
I encountered the same error trying to debug why I couldn't connect to BitBucket. I initially increased the size of the 'fread' buffer from 1024 to 2048 (Server.php readClientRequest:153)

```
private function readClientRequest($clientSocket)
    {
        $request = '';
        while (($chunk = @fread($clientSocket, 2048)) !== false) {
            $request .= $chunk;
            if (strpos($request, "\r\n\r\n") !== false) {
                break;
            }
        } ...
```
which gave me some relief but the truncation occurred again. Turns out that the buffer is filling up due to me running the deploy script multiple times trying to figure out my connection issue. The workaround for me was to delete the .dep/releases.log file in the deployment directory to reset the buffer. I hope this helps.

**Reply by pelmered-grebban** [reply]
> I encountered the same error trying to debug why I couldn't connect to BitBucket. I initially increased the size of the 'fread' buffer from 1024 to 2048 (Server.php readClientRequest:153)
> 
> ```
> private function readClientRequest($clientSocket)
>     {
>         $request = '';
>         while (($chunk = @fread($clientSocket, 2048)) !== false) {
>             $request .= $chunk;
>             if (strpos($request, "\r\n\r\n") !== false) {
>                 break;
>             }
>         } ...
> ```
> 
> which gave me some relief but the truncation occurred again. Turns out that the buffer is filling up due to me running the deploy script multiple times trying to figure out my connection issue. The workaround for me was to delete the .dep/releases.log file in the deployment directory to reset the buffer. I hope this helps.

This worked for me as well, but I had to bump it up even higher to fint the whole thing in one chunk. 16384 has worked for me so far.

**Reply by antonmedv** [reply]
But why is the while loop not working? It should read the whole thing. I guess.

**Reply by NickMous** [reply]
It cuts the buffer off at 1024 bytes, and checks if we have 2 newlines with the `strpos` function as seen in the code snippet above. Somehow the headers contain one empty line exactly at the 1024 bytes. So if I'm not the only one maybe crank it up a bit or a different number in it for now? I have been extremely busy so not been able to create a good fix sadly.

**Reply by antonmedv** [reply]
Ufff. This is very strange behavior. Could you please share what JSON object is transferred? I may try to debug it and find current solution. Deployer v8-alpha, right?

**Reply by underdpt** [reply]
Hello,

I've been also hit by this, on v8-alpha, trying to provision a new server. This is the json object being transferred (it's been anonymized, but maintaining the character count):

```json
{
    "alias": "mercado-XXXXXXXX.mainsip.pro",
    "hostname": "mercado-XXXXXXXX.mainsip.pro",
    "stage": "production",
    "branch": "master",
    "remote_user": "ubuntu",
    "deploy_path": "\/var\/www\/mercado.XXXXXXXXXXXXXX.com",
    "keep_releases": 5,
    "sudo_password": "",
    "domain": "mercado-XXXXXXXX.mainsip.pro",
    "public_path": "public",
    "php_version": "8.4",
    "db_type": "mysql",
    "db_user": "XXXXXXXX",
    "db_name": "XXXXXXX",
    "db_password": "XXXXXXXXXXXXXXXX",
    "selected_hosts": [
        "mercado-XXXXXXXX.mainsip.pro"
    ]
}
```

**Comment by zjalen** [comment]
I encountered the same problem. On the Windows 10 platform, version 8.0.0-alpha.

**Comment by pelmered** [comment]
I fixed this in a PR: https://github.com/deployphp/deployer/pull/4032

I changed the `readClientRequest` method to this:

```php
    private function readClientRequest($clientSocket)
    {
        $request = stream_get_contents($clientSocket);

        if ($request === false) {
            throw new Exception('Socket read failed');
        }

        return $request;
    }
```

~~Feel free to use my fork while we wait for this to be fixed. Then you need to update your deployer version in `composer.json` to `"deployer/deployer": "dev-master as 8.0",` and add the following repository:~~
```
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/pelmered/deployer.git"
        }
    ],
```
Then run `composer update deployer/deployer`.

**Reply by pelmered** [reply, marked as answer]
It is already merged now. so now it is enough with setting the version to `dev-master as 8.0` and skip the repositories to use my fork.
